File: c++/0046.cpp
// Link: https://leetcode.com/problems/permutations/description/
// Runtime: 0 ms
// Memory: 8.33 MB

class Solution 
{
public:
    std::vector<vector<int>> permute(const std::vector<int>& nums) 
    {
        addPermutations(nums, {});
        return permutations;
    }

private:
    std::vector<std::vector<int>> permutations;

    void addPermutations(const std::vector<int>& nums, std::vector<int> permutation)
    {
        if (static_cast<int>(permutation.size()) == static_cast<int>(nums.size())) permutations.push_back(permutation);
        else
        {
            for (const auto& val : nums)
            {
                if (std::find(permutation.begin(), permutation.end(), val) == permutation.end())
                {
                    permutation.push_back(val);
                    addPermutations(nums, permutation);
                    permutation.pop_back();
                }
            }
        }
    }
};

File: python/0046.py
# Link: https://leetcode.com/problems/permutations/description/
# The first solution
# Runtime: 51 ms
# Memory: 16.5 MB


class Solution:
    permutations = []

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.permutations = []
        self.__add_permutations(nums, [])

        return self.permutations

    def __add_permutations(self, nums: List[int], permutation: List[int]) -> None:
        if len(permutation) == len(nums):
            self.permutations.append(permutation[:])
        else:
            for _, value in enumerate(nums):
                if value not in permutation:
                    permutation.append(value)
                    self.__add_permutations(nums, permutation)
                    permutation.pop()


# The second solution
# Runtime: 34 ms
# Memory: 16.3 MB


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        return [permutation for permutation in itertools.permutations(nums)]
